### ISSUE TITLE
callback error must be non literal?

### ISSUE THREAD

**Issue by xjamundx** [xjamundx/eslint-plugin-standard#12]
I was wondering if we could have `no-callback-literal` rule to correspond with the `no-throw-literal` rule.

Basically, if you're using the callback pattern this should never be okay:

``` js
callback('error')
```

instead you should do something more like this:

``` js
callback(new Error('error'))
```

We already built this locally, so let me know :)

See: https://github.com/eslint/eslint/issues/6427

**Comment by feross** [comment]
Yeah, I think this is a good idea. I like your list of `callback`, `cb`, and `next`. However, there's definitely a chance that there might be lots of false positives, but let's give it a try!

If you send a PR, I can test it with the test suite and see how many repos start failing -- and how many of those are valid vs. invalid.

**Comment by xjamundx** [comment]
Oh yeah let me do the PR today

**Comment by xjamundx** [comment]
Will get back to this today :)

**Comment by feross** [comment]
Nice :)

**Comment by xjamundx** [comment]
Here's what we're using at work....Will need to make it a bit more configurable. PR will come once I actually start working this morning...

```js
'use strict';

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

function isCallback(name) {
  return ['callback', 'next', 'cb'].indexOf(name) > -1
}

/**
 * Determine if a node has a possiblity to be an Error object
 * @param  {ASTNode}  node  ASTNode to check
 * @returns {boolean}       True if there is a chance it contains an Error obj
 */
function couldBeError(node) {
  switch (node.type) {
    case 'Identifier':
    case 'CallExpression':
    case 'NewExpression':
    case 'MemberExpression':
    case 'TaggedTemplateExpression':
    case 'YieldExpression':
      return true; // possibly an error object.

    case 'AssignmentExpression':
      return couldBeError(node.right);

    case 'SequenceExpression':
      var exprs = node.expressions;

      return exprs.length !== 0 && couldBeError(exprs[exprs.length - 1]);

    case 'LogicalExpression':
      return couldBeError(node.left) || couldBeError(node.right);

    case 'ConditionalExpression':
      return couldBeError(node.consequent) || couldBeError(node.alternate);

    default:
      return node.value === null
  }
}

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = function(context) {

    return {

      CallExpression: function(node) {
        var errorArg = node.arguments[0]
        if (errorArg && isCallback(node.callee.name)) {
          if (!couldBeError(errorArg)) {
            context.report(node, 'Unexpected literal in error position of callback.');
          } else if (node.arguments.length > 1 && errorArg.type === 'Identifier') {
            if (errorArg.name === 'undefined') {
              context.report(node, 'Expected null instead found undefined.');
            }
          }
        }
      }

    };
};
```

**Comment by pakastin** [comment]
What's wrong with using:
```js
if (err) {
  console.error(err);
  return next('Something went wrong');
}
```
In Express.js error handling works like that: https://expressjs.com/en/guide/error-handling.html

**Comment by xjamundx** [comment]
@pakastin I think it should work as long as you're not passing a string

**Comment by pakastin** [comment]
Oh, true – I meant that if you want to customize error messages:
```
if (err) {
  console.error(err);
  return next('DB error: user');
}
```

**Comment by xjamundx** [comment]
@pakastin so the way this rule would suggest doing that is:

```js
if (err) {
  console.error(err);
  return next(new Error('DB error: user'));
}
```

The benefit is that using an error object, will keep the message (as the `message` property, but it will *also* give you `stack` property so you can see exactly where the error was called!)

**Comment by pakastin** [comment]
Oh, that's nice!

**Comment by jonashaag** [comment]
This just hit me in a non-error case. There are definitely valid use cases where you pass a string that's not an error. Sorry to say that, but this rule is just stupid.

**Comment by jonashaag** [comment]
This also seems to match

```
callback({foo: 'bar'})
```

which *definitely* is prone to lots of false positives.

**Comment by jonashaag** [comment]
Guys, this also matches

```
callback({foo: somevariable})
```

Stop it already :-D

**Comment by TheAdamizer** [comment]
Ran into this rule while writing some tests and passing in a couple numbers into a parameter called 'callback.'  Got around it by renaming the parameter to aFunc.  There's no real pattern matching going on here, just looking for some parameters of certain names it seems (which doesn't seem like a worthwhile use of a linter to me).  Probably worth at least revisiting this test...

**Comment by jessereitsma** [comment]
```
.catch(() => callback(false)); // eslint-disable-line standard/no-callback-literal
``` 
The line comment saved me

**Comment by timkinnane** [comment]
I got this lint error in an unrelated context, I think just from using an argument named callback...

In my case, nothing to do with error handling, I've got an es6 class listening to and relaying events on another module with a custom `on` function:

```
on (event, callback) {
    otherModule.on(event, (...args) => {
      callback(...args)
    })
  }
```

**Comment by xjamundx** [comment]
We should probably make an exception for the spread operator!

**Comment by jonashaag** [comment]
> We should probably make an exception for the spread operator!

You should drop the rule entirely guys. This simply has too many false positives.

**Comment by hiimbex** [comment]
Agree with what's been said about this rule being silly. Passing a string is valid in a lot of use cases, especially for testing.

**Comment by xjamundx** [comment]
@feross I'm basically fine removing this rule, I've personally moved to async/await instead of callbacks wherever possible and it's clear that a lot of people don't want to follow the callback-pattern strictly.

**Comment by feross** [comment]
Opened an issue to discuss removing this rule: https://github.com/standard/eslint-plugin-standard/issues/27
